### ISSUE TITLE
[TTree] Inconcistency with the TTreeReader's SetEntries.

### ISSUE THREAD

**Issue by ikabadzhov** [root-project/root#10774]
I encountered the following inconcistent behaviour in the `TTreeReader`
```cpp
   auto dfWriter0 = RDataFrame(3).Define("x", [] { return 1; });
   dfWriter0.Snapshot<int>("treeA", "file0.root", {"x"});
   auto chain = std::make_shared<TChain>("treeA");
   chain->Add("file0.root");
   TTreeReader r(chain.get(), chain->GetEntryList());
   r.SetEntriesRange(4, 5);
```
this produces: `Error in <TTreeReader::SetEntriesRange()>: Error setting first entry 4: one of the readers was not successfully initialized`, which is good (as expected).

On the other hand, if the range becomes `r.SetEntriesRange(3, 5)` - start is now the last index + 1, there is no error produced or etc.

Using gdb, I saw that there is a call in `SetEntriesRange` to `SetEntry(beginEntry - 1)` (https://github.com/root-project/root/blob/f841b96e634b7f7136b0c80da0f9394e8e370314/tree/treeplayer/src/TTreeReader.cxx#L438)

`SetEntry` however, calls `SetEntryBase`, which modifies `fEntry`. Hence, the quickfix to call `SetEntry(beginEntry)` is obviously dangerous.

CC: @eguiraud @vepadulano

**Comment by eguiraud** [comment]
I think the underlying problem is that `SetEntriesRange` has no input validation. We could (and should) error out at the point `SetEntriesRange` is called if `start > nEntries`.

**Comment by ikabadzhov** [comment]
I can take care of making a check before calling SetEntriesRange. I was more concerned whether SetEntriesRange can become dangerous on its own.

**Comment by eguiraud** [comment]
> SetEntriesRange can become dangerous on its own

what do you mean?

**Comment by ikabadzhov** [comment]
> > SetEntriesRange can become dangerous on its own
> 
> what do you mean?

User not getting notified for specifying invalid range (e.g. start=last entry)

**Comment by eguiraud** [comment]
Yes that's what input validation is for, what am I missing?

**Comment by ikabadzhov** [comment]
Fyi I am currently doing my own sanity check (outside `SetEntriesRange`) by first calling `SetEntry` with the desired beginning. 

`SetEntry` returns an entry status, which is 0 upon success, and works in my case as expected, i.e. it properly catches that `SetEntry(lastEntry)` is invalid.

**Comment by eguiraud** [comment]
would it be possible to move that same check inside `SetEntriesRange`?

**Comment by ikabadzhov** [comment]
To my understanding, yes. Indeed, there is a call to `GetEntries(false)` here: https://github.com/root-project/root/blob/f841b96e634b7f7136b0c80da0f9394e8e370314/tree/treeplayer/src/TTreeReader.cxx#L418

An argument why is not done with `GetEntries(true)` is ofc efficiency - as at this stage, there is no need to open all files in case of a chain.

However, within `SetEntriesRange` as previously mentioned, there is also a call to `SetEntry`, here: https://github.com/root-project/root/blob/f841b96e634b7f7136b0c80da0f9394e8e370314/tree/treeplayer/src/TTreeReader.cxx#L438. Which will do all of the work of obtaining all entry nummbers.

**Comment by ikabadzhov** [comment]
See:
```
[0] from 0x00007ffff755d578 in TFile::Open(char const*, char const*, char const*, int, int)+54 at [...]/io/io/src/TFile.cxx:4024
[1] from 0x00007ffff66a5cab in TChain::LoadTree(long long)+3431 at [...]/tree/tree/src/TChain.cxx:1499
[2] from 0x00007ffff647a69f in TTreeReader::SetEntryBase(long long, bool)+653 at [...]/tree/treeplayer/src/TTreeReader.cxx:547
[3] from 0x00007ffff647b132 in TTreeReader::SetEntry(long long)+40 at [...]/tree/treeplayer/inc/TTreeReader.h:199
[4] from 0x00007ffff647a154 in TTreeReader::SetEntriesRange(long long, long long)+322 at [...]/tree/treeplayer/src/TTreeReader.cxx:438
```
and then it continues:
```
[0] from 0x00007ffff755d578 in TFile::Open(char const*, char const*, char const*, int, int)+54 at [...]/io/io/src/TFile.cxx:4024
[1] from 0x00007ffff66a5cab in TChain::LoadTree(long long)+3431 at [...]/tree/tree/src/TChain.cxx:1499
[2] from 0x00007ffff66a635f in TChain::LoadTree(long long)+5147 at [...]/tree/tree/src/TChain.cxx:1586
[3] from 0x00007ffff647a69f in TTreeReader::SetEntryBase(long long, bool)+653 at [...]/tree/treeplayer/src/TTreeReader.cxx:547
[4] from 0x00007ffff647b132 in TTreeReader::SetEntry(long long)+40 at [...]/tree/treeplayer/inc/TTreeReader.h:199
[5] from 0x00007ffff647a154 in TTreeReader::SetEntriesRange(long long, long long)+322 at [...]/tree/treeplayer/src/TTreeReader.cxx:438
```

**Comment by eguiraud** [comment]
That's not opening _all_ files, just the one that contains the entry that needs loading

**Comment by ikabadzhov** [comment]
Agreed. But the issue that specifying the last index + 1 as a start is still there! Hence this PR should remain open imo.

**Comment by ikabadzhov** [comment]
Yes! Minor remark: not a file, but a chain of multiple trees/files.

For a single file, it works as expected.
